In [1]:
#import necessary library
import numpy as np # linear algebra
import pandas as pd # data manipulation and analysis
import matplotlib.pyplot as plt # data visualization
import seaborn as sns # data visualization
sns.set_style('whitegrid') # set style for visualization
import warnings # ignore warnings
warnings.filterwarnings('ignore')
from initial_report import *

plt.style.use("ggplot")
from matplotlib.pyplot import figure


In [2]:
#read csv file
df_customer=pd.read_csv("customers.csv")

In [3]:
#read sample data
df_customer.head(5)

customer_id  became_member_on gender  age    income
0  68be06ca386d4c31939f3a4f0e3dd783          20170212    NaN  118       NaN
1  0610b486422d4921ae7d2bf64640c50b          20170715      F   55  112000.0
2  38fe809add3b4fcf9315a9694bb96ff5          20180712    NaN  118       NaN
3  78afa995795e4d85b5d9ceeca43f5fef          20170509      F   75  100000.0
4  a03223e636434f42ac4c3df47e8bac43          20170804    NaN  118       NaN

1. gender and income has null value
2. date column is not in proper data type

In [4]:
#check initial report
initial_report(df_customer)

 *** DATA CLEANING CHECKLIST ***
----------------------------------------
*** Structure:
- Total Rows: 17000
- Total Columns: 5
- Column Names: ['customer_id', 'became_member_on', 'gender', 'age', 'income']

📌 Data Types:
  customer_id: object
  became_member_on: int64
  gender: object
  age: int64
  income: float64

🧬 Mixed Data Types:
  gender:
    - str: 14825
    - float: 2175

*** Distinct Values per Column:
  customer_id: 17000
  became_member_on: 1716
  gender: 3
  age: 85
  income: 91

*** Null Values and Percentages:
  gender: Missing Values: 2175, Pct: 12.794%
  income: Missing Values: 2175, Pct: 12.794%


*** Duplicates: 0

*** Negative or Zero Values:

*** Basic Statistics:
       became_member_on           age         income
count      1.700000e+04  17000.000000   14825.000000
mean       2.016703e+07     62.531412   65404.991568
std        1.167750e+04     26.738580   21598.299410
min        2.013073e+07     18.000000   30000.000000
25%        2.016053e+07     45.000000   

1. gender column has mixed data type
2. null values: gender [2175, 12.794%], income [ 2175, 12.794%]
3. age column has outlier

In [5]:
#unique list from gender columns
list(df_customer.gender.unique())

[nan, 'F', 'M', 'O']

In [6]:
#value counts from gender columns
df_customer.gender.value_counts()

gender
M    8484
F    6129
O     212
Name: count, dtype: int64

In [7]:
#became_member_on date column fixing
df_customer['became_member_on'] = pd.to_datetime(df_customer['became_member_on'], format='%Y%m%d')

In [8]:
#check sample
df_customer.sample(5)

customer_id became_member_on gender  age   income
12020  79b174cbc5004c84a5ade3eab1aa409a       2015-08-16      M   41  60000.0
7916   92105bd31e38472ab74fbb1345e0ad7b       2017-09-28      M   45  92000.0
15180  86b8f92aa604408994acf65809e10f7f       2017-02-21    NaN  118      NaN
7673   adcbfd37ea65430d852cb783808d05da       2017-08-21      M   56  46000.0
2188   9885e229fb8541afb0fb5d4d25eef01f       2017-12-12      M   43  75000.0

In [9]:
#check info
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customer_id       17000 non-null  object        
 1   became_member_on  17000 non-null  datetime64[ns]
 2   gender            14825 non-null  object        
 3   age               17000 non-null  int64         
 4   income            14825 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(2)
memory usage: 664.2+ KB


In [10]:
# Filter rows where gender is of float type
float_gender_rows = df_customer[df_customer['gender'].apply(lambda x: isinstance(x, float))]

# Display the result
print(float_gender_rows)

                            customer_id became_member_on gender  age  income
0      68be06ca386d4c31939f3a4f0e3dd783       2017-02-12    NaN  118     NaN
2      38fe809add3b4fcf9315a9694bb96ff5       2018-07-12    NaN  118     NaN
4      a03223e636434f42ac4c3df47e8bac43       2017-08-04    NaN  118     NaN
6      8ec6ce2a7e7949b1bf142def7d0e0586       2017-09-25    NaN  118     NaN
7      68617ca6246f4fbc85e91a2a49552598       2017-10-02    NaN  118     NaN
...                                 ...              ...    ...  ...     ...
16980  5c686d09ca4d475a8f750f2ba07e0440       2016-09-01    NaN  118     NaN
16982  d9ca82f550ac4ee58b6299cf1e5c824a       2016-04-15    NaN  118     NaN
16989  ca45ee1883624304bac1e4c8a114f045       2018-03-05    NaN  118     NaN
16991  a9a20fa8b5504360beb4e7c8712f8306       2016-01-16    NaN  118     NaN
16994  c02b10e8752c4d8e9b73f918558531f7       2015-12-11    NaN  118     NaN

[2175 rows x 5 columns]


1. float type showing null values

In [11]:
# Investigate Rows with Missing Values if there is any pattern
df_customer[df_customer['gender'].isnull()]

customer_id became_member_on gender  age  income
0      68be06ca386d4c31939f3a4f0e3dd783       2017-02-12    NaN  118     NaN
2      38fe809add3b4fcf9315a9694bb96ff5       2018-07-12    NaN  118     NaN
4      a03223e636434f42ac4c3df47e8bac43       2017-08-04    NaN  118     NaN
6      8ec6ce2a7e7949b1bf142def7d0e0586       2017-09-25    NaN  118     NaN
7      68617ca6246f4fbc85e91a2a49552598       2017-10-02    NaN  118     NaN
...                                 ...              ...    ...  ...     ...
16980  5c686d09ca4d475a8f750f2ba07e0440       2016-09-01    NaN  118     NaN
16982  d9ca82f550ac4ee58b6299cf1e5c824a       2016-04-15    NaN  118     NaN
16989  ca45ee1883624304bac1e4c8a114f045       2018-03-05    NaN  118     NaN
16991  a9a20fa8b5504360beb4e7c8712f8306       2016-01-16    NaN  118     NaN
16994  c02b10e8752c4d8e9b73f918558531f7       2015-12-11    NaN  118     NaN

[2175 rows x 5 columns]

1. where gender in NaN, age is 118, income is also NaN, verify again

In [12]:
# summary statistics for rows where the gender value is missing (NaN).
df_customer[df_customer['gender'].isnull()][['age', 'income']].describe()

age  income
count  2175.0     0.0
mean    118.0     NaN
std       0.0     NaN
min     118.0     NaN
25%     118.0     NaN
50%     118.0     NaN
75%     118.0     NaN
max     118.0     NaN

1. age is always showing 118. So we can delete these rows

In [13]:
#drop missing values
df_customer.dropna(subset=['gender'], inplace=True)

In [14]:
#check initial report now
initial_report(df_customer)

 *** DATA CLEANING CHECKLIST ***
----------------------------------------
*** Structure:
- Total Rows: 14825
- Total Columns: 5
- Column Names: ['customer_id', 'became_member_on', 'gender', 'age', 'income']

📌 Data Types:
  customer_id: object
  became_member_on: datetime64[ns]
  gender: object
  age: int64
  income: float64

🧬 Mixed Data Types:

*** Distinct Values per Column:
  customer_id: 14825
  became_member_on: 1707
  gender: 3
  age: 84
  income: 91

*** Null Values and Percentages:


*** Duplicates: 0

*** Negative or Zero Values:

*** Basic Statistics:
                    became_member_on           age         income
count                          14825  14825.000000   14825.000000
mean   2017-02-18 12:30:15.419898880     54.393524   65404.991568
min              2013-07-29 00:00:00     18.000000   30000.000000
25%              2016-05-20 00:00:00     42.000000   49000.000000
50%              2017-08-02 00:00:00     55.000000   64000.000000
75%              2017-12-30 00:00:0

In [15]:
#save cleaned files
df_customer.to_csv('cleaned_custoer_data.csv', index=False)